tests/asts.go

```diff
@@ -0,0 +1,181 @@
+			replicas:    0,
+			deleteSlots: sets.NewInt32(),
+		},
+		{
+			name:        "Scaling out tidb from 0 to 2 by adding pods 0 and 2",
+			component:   "tidb",
+			replicas:    2,
+			deleteSlots: sets.NewInt32(1),
+		},
+		{
+			name:        "Scaling tidb from 2 to 4 by deleting pods 2 and adding pods 3, 4 and 5",
+			component:   "tidb",
+			replicas:    4,
+			deleteSlots: sets.NewInt32(1, 2),
+		},
+		{
+			name:        "Scaling out pd from 3 to 5 by adding pods 3, 4",
+			component:   "pd",
+			replicas:    5,
+			deleteSlots: sets.NewInt32(),
+		},
+		{
+			name:        "Scaling in pd from 5 to 3 by deleting pods 0 and 3",
+			component:   "pd",
+			replicas:    3,
+			deleteSlots: sets.NewInt32(0, 3),
+		},
+		{
+			name:        "Scaling out pd from 3 to 5 by adding pods 5 and 6",
+			component:   "pd",
+			replicas:    5,
+			deleteSlots: sets.NewInt32(0, 3),
+		},
+	}
+}
+
+func AstsScalingTest(st AstsCase, clusterName, ns string, c, hc clientset.Interface, cli versioned.Interface, genericCli client.Client) {
+	ginkgo.By(st.name)
+	replicas := st.replicas
+	stsName := fmt.Sprintf("%s-%s", clusterName, st.component)
+
+	sts, err := hc.AppsV1().StatefulSets(ns).Get(stsName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	oldPodList := e2esset.GetPodList(c, sts)
+
+	ginkgo.By(fmt.Sprintf("Scaling sts %s/%s to replicas %d and setting deleting pods to %v (old replicas: %d, old delete slots: %v)", ns, stsName, replicas, st.deleteSlots.List(), *sts.Spec.Replicas, helper.GetDeleteSlots(sts).List()))
+	tc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(clusterName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+		if tc.Annotations == nil {
+			tc.Annotations = map[string]string{}
+		}
+		if st.component == "tikv" {
+			tc.Annotations[label.AnnTiKVDeleteSlots] = mustToString(st.deleteSlots)
+			tc.Spec.TiKV.Replicas = replicas
+		} else if st.component == "pd" {
+			tc.Annotations[label.AnnPDDeleteSlots] = mustToString(st.deleteSlots)
+			tc.Spec.PD.Replicas = replicas
+		} else if st.component == "tidb" {
+			tc.Annotations[label.AnnTiDBDeleteSlots] = mustToString(st.deleteSlots)
+			tc.Spec.TiDB.Replicas = replicas
+		} else {
+			return fmt.Errorf("unsupported component: %v", st.component)
+		}
+		return nil
+	})
+	framework.ExpectNoError(err)
+
+	ginkgo.By(fmt.Sprintf("Waiting for all pods of tidb cluster component %s (sts: %s/%s) are in desired state (replicas: %d, delete slots: %v)", st.component, ns, stsName, st.replicas, st.deleteSlots.List()))
+	err = wait.PollImmediate(time.Second*5, time.Minute*20, func() (bool, error) {
+		// check replicas and delete slots are synced
+		sts, err = hc.AppsV1().StatefulSets(ns).Get(stsName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+		if *sts.Spec.Replicas != st.replicas {
+			klog.Infof("replicas of sts %s/%s is %d, expects %d", ns, stsName, sts.Spec.Replicas, st.replicas)
```

Comment:
replace `sts.Spec.Replicas` with `*sts.Spec.Replicas`. This is a mistake I made.

Reply:
removed